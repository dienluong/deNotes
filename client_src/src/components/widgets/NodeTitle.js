import React from 'react';

function NodeTitle({ node, path, onSubmit: submitHandler }) {
  function submit(event) {
    const inputEl = event.target.matches('input') ? event.target : event.target.getElementsByTagName('input')[0];
    if (inputEl.value !== inputEl.defaultValue) {
      submitHandler({
        title: inputEl.value,
        node: node,
        path: path,
      });
    }
    event.preventDefault();
  }

  return (
    <form onSubmit={ submit } onBlur={ submit }>
      <input
        className='node-title'
        type="text"
        defaultValue={ node.title }
      />
    </form>
  );
}

export default NodeTitle;
