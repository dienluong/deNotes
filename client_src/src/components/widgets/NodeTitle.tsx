import React from 'react';
import styles from './NodeTitle.module.css';
import escape from 'lodash/escape';
import unescape from 'lodash/unescape';

// Types
import { FormEvent } from 'react';
type PropsT<N,P> = {
  node: N,
  path: P,
  onSubmit: (params: { title: string, node: N, path: P }) => unknown,
};

function NodeTitle({ node, path, onSubmit: submitHandler }: PropsT<TreeNodeT, any>) {
  const inputRef = React.createRef<HTMLInputElement>();
  function submit(event: FormEvent<HTMLElement>) {
    const inputEl = inputRef.current;
    if (inputEl instanceof HTMLInputElement) {
      inputEl.blur();
      const escapedInput = escape(inputEl.value);
      if (escapedInput !== inputEl.defaultValue) {
        submitHandler({
          title: escapedInput,
          node: node,
          path: path,
        });
      }
    }
    event.preventDefault();
  }

  return (
    <form onSubmit={ submit } >
      <input
        ref={ inputRef }
        className={ styles['dnt__node-title'] }
        type="text"
        defaultValue={ unescape(node.title) }
        onBlur={ submit }
        onClick={ event => event.stopPropagation() }
        onDoubleClick={ event => event.stopPropagation() }
      />
    </form>
  );
}

export default NodeTitle;
