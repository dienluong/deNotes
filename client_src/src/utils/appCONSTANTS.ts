interface nodeTypesT {
  FOLDER: Readonly<string>;
  ITEM: Readonly<string>;
}
export const nodeTypes: Readonly<nodeTypesT> = {
  FOLDER: 'FOLDER_NODE',
  ITEM: 'ITEM_NODE',
};

export const NONE_SELECTED: Readonly<string> = '<NO_NODE_SELECTED>';
