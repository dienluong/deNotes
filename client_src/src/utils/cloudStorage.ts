import Widget from 'remotestorage-widget';
import RemoteStorage from 'remotestoragejs';

// Types
type RemoteStorageT = typeof RemoteStorage;

const options = {
  leaveOpen: false,
  autoCloseAfter: 1500,
  skipInitial: true,
  logging: false,
};

export function connect({ storage }: { storage: RemoteStorageT  }) {
  if (storage) {
    const widget = new Widget(storage, options);
    storage.setApiKeys({
      dropbox: process.env.REACT_APP_DROPBOX_APPKEY,
    });
    return {
      storage,
      widget,
    }
  } else {
    return null;
  }
}
