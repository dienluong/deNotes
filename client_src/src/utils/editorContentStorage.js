let _save = () => Promise.reject(new Error('Save aborted. Cause: No Storage implementation provided.'));
let _load = () => Promise.reject(new Error('Load aborted. Cause: No Storage implementation provided.'));
let _remove = () => Promise.reject(new Error('Delete aborted. Cause: No Storage implementation provided.'));

export function inject({ save = _save, load = _load, remove = _remove }) {
  _save = typeof save === 'function' ? save : _save;
  _load = typeof load === 'function' ? load : _load;
  _remove = typeof remove === 'function' ? remove : _remove;
}

export function save({ userId, editorContent }) {
  if (!editorContent || !editorContent.content || !editorContent.delta || !editorContent.id) {
    return Promise.reject(new Error('Save aborted. Cause: invalid content.'));
  }

  if (!userId) {
    return Promise.reject(new Error('Save aborted. Cause: invalid userId.'));
  }

  return _save({
    collectionName: 'notes',
    id: editorContent.id,
    ownerId: userId,
    // TODO: Replace hardcoded values
    dataObj: {
      'id': editorContent.id,
      'title': editorContent.title || '',
      'body': editorContent.content,
      'delta': editorContent.delta,
      'dateCreated': editorContent.dateCreated,
      'dateModified': editorContent.dateModified,
      'ownerId': userId,
    },
  });
}

/**
 * Returns a promise for an array of content
 * @param noteId
 * @return {Promise<Response | never>}
 */
export function load({ id = '', userId = '' }) {
  if (!id && !userId) {
    return Promise.reject(new Error('Load aborted. Cause: invalid parameters.'));
  }

  return _load({
    collectionName: 'notes',
    id,
    ownerId: userId,
  }).then(content => {
    const propList = ['id', 'title', 'body', 'delta', 'dateCreated', 'dateModified'];
    if (content && (typeof content.hasOwnProperty === 'function') && propList.every(prop => content.hasOwnProperty(prop))) {
      // Note: We convert the retreived dates to milliseconds since Unix epoch.
      return {
        id,
        title: content.title,
        content: content.body,
        delta: typeof content.delta === 'string' ? JSON.parse(content.delta) : content.delta,
        dateModified: new Date(content.dateModified).getTime(),
        dateCreated: new Date(content.dateCreated).getTime(),
      };
    } else {
      const message = `Unrecognized data fetched. ID: ${id}`;
      return Promise.reject(new Error(message));
    }
  });
}

export function remove({ ids = '', userId = '' }) {
  if (typeof ids === 'string' || (Array.isArray(ids) && ids.length)) {
    return _remove({
      collectionName: 'notes',
      ids,
      ownerId: userId,
    });
  } else {
    return Promise.reject(new Error('Delete aborted. Cause: invalid parameter.'));
  }
}

