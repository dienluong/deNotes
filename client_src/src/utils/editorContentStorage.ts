let _save: Function = () => Promise.reject(new Error('Save aborted. Cause: No Storage implementation provided.'));
let _load: Function = () => Promise.reject(new Error('Load aborted. Cause: No Storage implementation provided.'));
let _remove: Function = () => Promise.reject(new Error('Delete aborted. Cause: No Storage implementation provided.'));

export function inject({ save = _save, load = _load, remove = _remove }: { save: Function, load: Function, remove: Function })
  : void {
  _save = typeof save === 'function' ? save : _save;
  _load = typeof load === 'function' ? load : _load;
  _remove = typeof remove === 'function' ? remove : _remove;
}

export function save({ userId, editorContent }: { userId: string, editorContent: EditorContentT })
  : Promise<any> {
  if (!editorContent || !editorContent.content || !editorContent.delta || !editorContent.id) {
    return Promise.reject(new Error('invalid content.'));
  }

  if (!userId || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid userId.'));
  }

  return _save({
    collectionName: 'notes',
    id: editorContent.id,
    ownerId: userId,
    dataObj: {
      'id': editorContent.id,
      'title': editorContent.title || '',
      'body': editorContent.content,
      'delta': editorContent.delta,
      'dateCreated': editorContent.dateCreated || 0,
      'dateModified': editorContent.dateModified || 0,
      'ownerId': userId,
    },
  });
}

/**
 * Returns a promise for an array of content
 * @param noteId
 * @return {Promise<Response | never>}
 */
export function load({ id = '', userId = '' }: { id: string, userId: string })
  : Promise<EditorContentT> {
  if (!id || !userId || typeof id !== 'string' || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid parameters.'));
  }

  return _load({
    collectionName: 'notes',
    id,
    ownerId: userId,
  }).then((content: any) => {
    const propList = ['id', 'title', 'body', 'delta', 'dateCreated', 'dateModified'];
    if (content && (typeof content.hasOwnProperty === 'function') && propList.every(prop => content.hasOwnProperty(prop))) {
      // Note: We convert the retreived dates to milliseconds since Unix epoch.
      return {
        id,
        title: content.title,
        content: content.body,
        delta: typeof content.delta === 'string' ? JSON.parse(content.delta) : content.delta,
        dateModified: new Date(content.dateModified).getTime(),
        dateCreated: new Date(content.dateCreated).getTime(),
      };
    } else {
      const message = `Unrecognized data fetched. ID: ${id}`;
      return Promise.reject(new Error(message));
    }
  });
}

export function remove({ ids = '', userId = '' }: { ids: string, userId: string })
  : Promise<{ results: string|string[], count: number }> {
  if (!ids || !userId || (typeof ids !== 'string' && !Array.isArray(ids)) || (Array.isArray(ids) && !ids.length) || typeof userId !== 'string') {
    return Promise.reject(new Error('invalid parameters.'));
  } else {
    return _remove({
      collectionName: 'notes',
      ids,
      ownerId: userId,
    });
  }
}

