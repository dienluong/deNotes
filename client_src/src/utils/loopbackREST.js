/**
 * @param id
 * @param dataObj
 * @return {Promise<Response | never>}
 */
export function save(id, dataObj) {
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/trees/upsertWithWhere?where={"ownerId": ${id}}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(dataObj),
    }
  );
}

/**
 * @param id
 * @return {Promise<Response | never>}
 */
export function load(id) {
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/trees?where={"ownerId": ${id}}`,
    {
      method: 'GET',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
    },
  );
}
