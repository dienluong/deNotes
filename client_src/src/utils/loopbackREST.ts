/** Update matching record or create new one if none exists. If both id and ownerId provided, ownerId is ignored.
 * @param collectionName {string} Name of the table or collection
 * @param id {string}
 * @param ownerId {string}
 * @param dataObj
 * @return {Promise<Response | never>}
 */
export function save({ collectionName = '', id = '', ownerId = '', dataObj }
  : { collectionName: string, id: string, ownerId: string, dataObj: object })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Save aborted: invalid collection name.'));
  }
  if (!id && !ownerId) {
    return Promise.reject(new Error('Save aborted: no valid ID provided.'));
  }

  let filter = '';
  if (id) {
    filter = '?where=' + encodeURIComponent(`{"id":"${id}"}`);
  } else if (ownerId) {
    filter = '?where=' + encodeURIComponent(`{"ownerId":"${ownerId}"}`);
  }

  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}/upsertWithWhere${filter}`,
    {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'application/json',
      },
      body: JSON.stringify(dataObj),
    }
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR saving data to remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

/**
 * If multiple items match, results are returned in an array.
 * @param collectionName {string} Name of the table or collection
 * @param id {string}
 * @param ownerId {string}
 * @return {Promise<Response | never>}
 */
export function load({ collectionName = '', id = '', ownerId = '' }: { collectionName: string, id: string, ownerId: string })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Load aborted: invalid collection name.'));
  }
  const ownerFilter = ownerId ? '?where=' + encodeURIComponent(`{"ownerId": "${ownerId}"}`) : '';
  const idFilter = id ? `/${id}` : '';
  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}${idFilter}${ownerFilter}`,
    {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
      },
    },
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR loading data from remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

export function remove({ collectionName = '', ids = '' }: { collectionName: string, ids: string|string[] })
  : Promise<any> {
  if (!collectionName) {
    return Promise.reject(new Error('Delete aborted: invalid collection name.'));
  }

  let idFilter;
  if (ids && typeof ids === 'string') {
    idFilter = `/${ids}`;
  } else if (Array.isArray(ids) && ids.length) {
    idFilter = `/multidelete?ids=${encodeURIComponent(JSON.stringify(ids))}`;
  } else {
    return Promise.reject(new Error('Delete aborted: invalid filter'));
  }

  return fetch(
    `${process.env.REACT_APP_SERVER_URL}/api/${collectionName}${idFilter}`,
    {
      method: 'DELETE',
      headers: {
        'Accept': 'application/json',
      },
    },
  ).then(response => {
    if (response.ok) {
      return response.json();
    } else {
      return Promise.reject(new Error(`ERROR deleting item from remote storage! ${response.status} - ${response.statusText}`));
    }
  });
}

