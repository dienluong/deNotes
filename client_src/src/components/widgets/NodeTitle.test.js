import NodeTitle from './NodeTitle';
import React from 'react';
import { render, cleanup, fireEvent } from 'react-testing-library';
import 'jest-dom/extend-expect';
import { nodeTypes } from '../../utils/appCONSTANTS';

afterEach(cleanup);

it('renders an element composed of an input with value corresponding to node title', () => {
  const props = {
    node: {
      title: 'test-node',
      id: 'test-node-id',
      type: nodeTypes.FOLDER,
      expanded: true,
      children: [],
    },
    path: ['step1', 'step2'],
    onSubmit: jest.fn(),
  };

  const { container } = render(<NodeTitle {...props} />);
  expect(container.getElementsByTagName('input')[0]).toHaveAttribute('value', props.node.title);
  expect(container).toMatchSnapshot();
});

it('calls handler function w/ the current input value on blur event, if value changed', () => {
  const props = {
    node: {
      title: 'test-node',
      id: 'test-node-id',
      type: nodeTypes.FOLDER,
      expanded: true,
      children: [],
    },
    path: ['step1', 'step2'],
    onSubmit: jest.fn(),
  };

  const newName = 'new folder name';

  const { container } = render(<NodeTitle {...props} />);
  const inputEl = container.getElementsByTagName('input')[0];
  inputEl.value = newName;
  fireEvent.blur(inputEl);
  expect(props.onSubmit).toBeCalledTimes(1);
  expect(props.onSubmit).lastCalledWith({ title: newName, node: props.node, path: props.path });
});

it('calls handler function w/ the current input value on submit event, if value changed', () => {
  const props = {
    node: {
      title: 'test-node',
      id: 'test-node-id',
      type: nodeTypes.FOLDER,
      expanded: true,
      children: [],
    },
    path: ['step1', 'step2'],
    onSubmit: jest.fn(),
  };

  const newName = 'different folder name';

  const { container } = render(<NodeTitle {...props} />);
  container.getElementsByTagName('input')[0].value = newName;
  const formEl = container.getElementsByTagName('form')[0];
  fireEvent.submit(formEl);
  expect(props.onSubmit).toBeCalledTimes(1);
  expect(props.onSubmit).lastCalledWith({ title: newName, node: props.node, path: props.path });
});

it('does not call handler function, if value did not change', () => {
  const props = {
    node: {
      title: 'test-node',
      id: 'test-node-id',
      type: nodeTypes.FOLDER,
      expanded: true,
      children: [],
    },
    path: ['step1', 'step2'],
    onSubmit: jest.fn(),
  };

  const { container } = render(<NodeTitle {...props} />);

  const formEl = container.getElementsByTagName('form')[0];
  fireEvent.submit(formEl);
  const inputEl = container.getElementsByTagName('input')[0];
  fireEvent.blur(inputEl);

  expect(props.onSubmit).not.toBeCalled();
});

it('does not propagate click event', () => {
  const props = {
    node: {
      title: 'test-node',
      id: 'test-node-id',
      type: nodeTypes.FOLDER,
      expanded: true,
      children: [],
    },
    path: ['step1', 'step2'],
    onSubmit: jest.fn(),
  };

  const onClickHandler = jest.fn();
  const { container } = render(<div onClick={ onClickHandler }><NodeTitle {...props} /></div>);
  const inputEl = container.getElementsByTagName('input')[0];
  fireEvent.click(inputEl);

  expect(onClickHandler).not.toBeCalled();
});
