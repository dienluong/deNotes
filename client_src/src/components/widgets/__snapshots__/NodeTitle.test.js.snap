// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an element composed of an input with value corresponding to node title 1`] = `
<div>
  <form>
    <input
      class="dnt__node-title"
      type="text"
      value="test-node"
    />
  </form>
</div>
`;
